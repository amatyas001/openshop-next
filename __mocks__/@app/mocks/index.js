export const mockProduct = (repeat = 1) => {
  let index = 0;
  let items = [];
  do {
    items.push({
      id: `mock_id_${index}`,
      amount: 1,
      color: `mock_color_${index}`,
      description: `mock_description_${index}`,
      gender: `mock_gender_${index}`,
      img: `mock_img_${index}`,
      name: `mock_name_${index}`,
      price: 1,
      review: `mock_review_${index}`,
      sizes: [`mock_size_0_${index}`, `mock_size_1_${index}`],
      starrating: 1,
    });
  } while (--repeat !== 0);
  return items;
};
