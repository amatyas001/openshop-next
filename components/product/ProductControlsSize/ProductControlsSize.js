import PropTypes from 'prop-types';
import { Text, Select } from '@chakra-ui/core';
import { FaArrowDown } from 'react-icons/fa';
import * as COLORS from '@app/config/colors';

const MemoProductControlsSize = (props) => {
  const { product, details, setDetails } = props;
  return product.sizes ? (
    <Select
      placeholder='Select size...'
      icon={FaArrowDown}
      height='50px'
      width='85%'
      fontSize='1.4rem'
      fontFamily='Khand'
      fontWeight='bold'
      mx='auto'
      bg={COLORS.BUTTON.transparent.base}
      color={COLORS.BUTTON.primary.base}
      border='0'
      borderRadius='0'
      value={details.buy.size || ''}
      onChange={(e) => setDetails(e.target.value)}
      _hover={{
        ...COLORS.BUTTON.transparent.hover,
        color: COLORS.BUTTON.primary.hover.bg,
        textDecoration: 'none',
        cursor: 'pointer',
      }}
      _focus={{
        ...COLORS.BUTTON.transparent.hover,
        color: COLORS.BUTTON.primary.hover.bg,
        textDecoration: 'none',
      }}
      _select={{
        ...COLORS.BUTTON.transparent.hover,
        color: COLORS.BUTTON.primary.hover.bg,
        textDecoration: 'none',
      }}
      _active={{
        ...COLORS.BUTTON.transparent.hover,
        color: COLORS.BUTTON.primary.hover.bg,
        textDecoration: 'none',
      }}
    >
      {product.sizes.map((item) => (
        <option key={item} value={item}>
          {item}
        </option>
      ))}
    </Select>
  ) : (
    <Text
      color={COLORS.BUTTON.primary.base}
      fontFamily='Khand'
      fontSize='1.4rem'
      fontWeight='bold'
      mx='auto'
    >
      One Size
    </Text>
  );
};

MemoProductControlsSize.propTypes = {
  /**
   * Unique product object
   */
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    color: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    description: PropTypes.string,
    gender: PropTypes.string,
    img: PropTypes.string,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    review: PropTypes.string,
    sizes: PropTypes.arrayOf(PropTypes.string),
    starrating: PropTypes.number.isRequired,
  }).isRequired,

  /**
   * Unique product object and its buying details
   */
  details: PropTypes.shape({
    id: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    color: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    description: PropTypes.string,
    gender: PropTypes.string,
    img: PropTypes.string,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    review: PropTypes.string,
    sizes: PropTypes.arrayOf(PropTypes.string),
    starrating: PropTypes.number.isRequired,
    buy: PropTypes.shape({
      amount: PropTypes.number.isRequired,
      size: PropTypes.string,
      color: PropTypes.string,
    }),
  }).isRequired,

  /**
   * Handler to set product buying details in wrapper
   */
  setDetails: PropTypes.func.isRequired,
};

/**
 * Renders a select input element with available product sizes as options
 * or plain *One Size* text if there are no sizes given.
 *
 * @see https://amatyas001.github.io/openshop-next/#productcard
 */
export const ProductControlsSize = React.memo(
  MemoProductControlsSize,
  /* istanbul ignore next */
  (prev, next) => prev.details.buy.size === next.details.buy.size
);
