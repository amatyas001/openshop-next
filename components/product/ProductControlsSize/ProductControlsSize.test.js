import { create, act } from 'react-test-renderer';
import { ProductControlsSize } from './ProductControlsSize';
import { mockProduct } from '@app/__mocks__/@app/mocks';

describe('<ProductControlsSize />', () => {
  let tree;

  const mock_product = {
    ...mockProduct(1)[0],
    sizes: ['mock_size_0', 'mock_size_1'],
  };

  const mock_details = {
    ...mock_product,
    buy: {
      amount: 0,
    },
  };

  const mock_setDetails = jest.fn();

  const mock_value = 'mock_value';

  beforeAll(() => {
    act(() => {
      tree = create(
        <ProductControlsSize
          product={mock_product}
          details={mock_details}
          setDetails={mock_setDetails}
        />
      );
    });
  });

  it('should render with required props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call handler', () => {
    act(() => {
      tree.root
        .findByType('select')
        .props.onChange({ target: { value: mock_value } });
    });
    expect(mock_setDetails).toHaveBeenCalledTimes(1);
    expect(mock_setDetails).toHaveBeenCalledWith(mock_value);
  });

  it('should render static value if no sizes available for the product', () => {
    act(() => {
      tree = create(
        <ProductControlsSize
          product={{ ...mock_product, sizes: undefined }}
          details={mock_details}
          setDetails={mock_setDetails}
        />
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
