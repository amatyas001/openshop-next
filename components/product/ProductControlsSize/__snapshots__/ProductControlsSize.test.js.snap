// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductControlsSize /> should render static value if no sizes available for the product 1`] = `
<p
  className="css-u84yc3"
>
  One Size
</p>
`;

exports[`<ProductControlsSize /> should render with required props 1`] = `
<div
  className="css-1php3x8"
>
  <select
    className="css-1fbi3l4"
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Select size...
    </option>
    <option
      value="mock_size_0"
    >
      mock_size_0
    </option>
    <option
      value="mock_size_1"
    >
      mock_size_1
    </option>
  </select>
  <div
    className="css-rl6gbg"
  >
    <svg
      aria-hidden="true"
      className="css-k176ff"
      data-custom-icon={true}
      fill="currentColor"
      focusable="false"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
      />
    </svg>
  </div>
</div>
`;
